Feature/hide a widget (#307)
MicroEJ/docs: ApplicationDeveloperGuide/UI/MWT/how-to-hide-widget.rst

Hiding a Widget
====================

> - To hide a widget, it's possible to use desktops to show and hide then at will 
> - in the sample below, The desktop will be exchanged with the one with the container/widget

Reviewer: ```suggestion
Hiding widgets can be done by:
- swapping the entire desktop; or
- selectively by placing your widgets in a custom hideable container.
```

> Using Desktop To hide a widget

Reviewer: please review the capitalisation rules

Reviewer 2: Still not good to me.

------------------------------

> In the code Sample below, a button is created to change the Desktop when clicked:

Reviewer: capitalisation
```suggestion
In the code sample below, a button is created to change the desktop when clicked:
```

.. code-block:: java

    MicroUI.start();
    Desktop desktop = new Desktop();

>     Canvas c;

Reviewer: use more meaningful variable names
```suggestion
    Canvas canvas;
```

>     Flow list = new Flow(Orientation");

Reviewer: why only a single quotation mark?

    list.addChild(new Label("Label 1"));
    list.addChild(new Label("Label 2"));

>     Button b = new Button("Click to change the Widget");

Reviewer: use more meaningful variable names

    Desktop desktop2 = new Desktop();
    b.setOnClickListener(new OnClickListener() {
        @Override
        public void onClick() {
            desktop2.setWidget(new Label("new Widget"));
            desktop2.requestShow();
        }
    });
    list.addChild(b);
    desktop.setWidget(list);
    desktop.requestShow();

> A new desktop with a widget is created, this will hide the Desktop, to go back to the first desktop you do the same process of calling requestShow from the desktop

Reviewer: ```suggestion
A new desktop with a single widget is created, hiding the original desktop. To go back just use requestShow on the first desktop.
```

.. code-block:: java

    MicroUI.start();
    Desktop desktop = new Desktop();
    Canvas c;
    Flow list = new Flow(LayoutOrientation.HORIZONTAL);
    list.addChild(new Label("Label 1"));
    list.addChild(new Label("Label 2"));
    Button b = new Button("Click to change the Widget");
    Desktop desktop2 = new Desktop();
    b.setOnClickListener(new OnClickListener() {
        @Override
        public void onClick() {
            Button b2 = new Button("Click to change back the Widget");
            b2.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick() {
                    desktop.requestShow();
                }
            });
            desktop2.setWidget(b2);
            desktop2.requestShow();
        }
    });
    list.addChild(b);
    desktop.setWidget(list);

>     desktop.requestShow();

Reviewer: duplicate code?

Author: There's a difference between them two

.. image:: images/hidewidgetshow.png

.. image:: images/hidewidgethidden.png

Using a Hideable Container to Hide a Widget
-------------------------------------------

> It is also possible to use a Hideable Container, which will render selectively the its children widget
> 
> On the widget hierarchy, a container is also a Widget, so if this Widget is a sibling of any other widget, There should be no Issues

Reviewer: annoyingly repetitive
```suggestion
This sample shows how to create a hideable container by disabling the rendering of its children.
```

.. code-block:: java
    
    	public class HideableContainer extends OverlapContainer {

		private boolean visibility = true;

		public boolean isVisible() {
			return this.visibility;
		}

		public void setVisible(boolean visible) {
			if (this.visibility != visible) {
				this.visibility = visible;

> 				this.setEnabled(visible);

Reviewer: explain why we are also disabling the children

Reviewer 2: I don't understand why it is also disabled. (Plus, I don't think it was enabled?)

How to Hide a Widget
====================

Hiding widgets can be done by:

- Swapping the entire desktop

> - Selectively by placing your widgets in a custom hideable container.

Reviewer 3: ```suggestion
- Selectively placing your widgets in a custom hideable container.
```
makes two "by"

> Using Desktop To hide a widget

Reviewer 2: I am not sure to understand how this section helps the user to "hide a widget". It creates a whole new desktop and when switching desktop ALL the widgets are hidden.

What could better illustrate the solution is to create exactly the same desktop (and the same widgets inside) but excluding one widget (and with the same layout).
BUT I don't think it is a good solution because it may cost a lot of Java heap.

> Using Desktop To hide a widget
> ------------------------------

Reviewer 3: ```suggestion
Using a Desktop to Hide a Widget
--------------------------------
```

In the code sample below, a button is created to change the desktop when clicked:

.. code-block:: java

    MicroUI.start();
    Desktop desktop = new Desktop();

>     Canvas canvas;

Reviewer 2: Useless.

>     Canvas canvas;

Reviewer 3: can be removed, never used!

>     Flow list = new Flow(LayoutOrientation.HORIZONTAL);

Reviewer 3: is it a flow or a list? maybe name it like `container` or so.

    list.addChild(new Label("Label 1"));
    list.addChild(new Label("Label 2"));

>     Button button = new Button("Click to change the Widget");

Reviewer 3: ```suggestion
    Button button = new Button("Click to hide the Widget");
```
hiding is the purpose of this article

    Desktop desktop2 = new Desktop();
    button.setOnClickListener(new OnClickListener() {
        @Override
        public void onClick() {

>             desktop2.setWidget(new Label("new Widget"));

Reviewer 3: This will not compile: 
`Cannot refer to the non-final local variable desktop2 defined in an enclosing scope`

Either:
- make `desktop2` `final`
- move `desktop2` to the `onClick()` method (preferred)

            desktop2.requestShow();
        }
    });
    list.addChild(button);
    desktop.setWidget(list);
    desktop.requestShow();

> A new desktop with a single widget is created, hiding the original desktop. To go back just use requestShow on the first desktop.

Reviewer 3: ### 
```suggestion
A new desktop with a single widget is created, replacing the original desktop. To go back just use `requestShow()` on the first desktop.
```

use `This is code` when refering to variables, methods, code etc.

.. code-block:: java

    MicroUI.start();
    Desktop desktop = new Desktop();

>     Canvas c;

Reviewer 2: Useless.

>     Canvas c;

Reviewer 3: Never used + avoid abbreviation when naming variables

    Flow list = new Flow(LayoutOrientation.HORIZONTAL);
    list.addChild(new Label("Label 1"));
    list.addChild(new Label("Label 2"));
    Button button = new Button("Click to change the Widget");
    Desktop desktop2 = new Desktop();
    button.setOnClickListener(new OnClickListener() {
        @Override
        public void onClick() {
            Button button2 = new Button("Click to change back the Widget");
            button2.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick() {

>                     desktop.requestShow();

Reviewer 3: This will not compile: 
`Cannot refer to the non-final local variable desktop defined in an enclosing scope`

make `desktop` `final`

                }
            });

>             desktop2.setWidget(b2);

Reviewer 2: ```suggestion
            desktop2.setWidget(button2);
```

>             desktop2.requestShow();

Reviewer 3: This will not compile: 
`Cannot refer to the non-final local variable desktop2 defined in an enclosing scope`

Either:
- make `desktop2` `final`
- move `desktop2` to the `onClick()` method (preferred)

        }
    });
    list.addChild(button);
    desktop.setWidget(list);
    desktop.requestShow();

.. image:: images/hidewidgetshow.png

.. image:: images/hidewidgethidden.png

Using a Hideable Container to Hide a Widget
-------------------------------------------

This sample shows how to create a hideable container by disabling the rendering of its children.

The children are not rendered, since the widget, and all of its children should be hidden.

.. code-block:: java
    

>     	public class HideableContainer extends OverlapContainer {

Reviewer 2: Why not directly adding this visible attribute to the button?

		private boolean visibility = true;

		public boolean isVisible() {
			return this.visibility;
		}

		public void setVisible(boolean visible) {
			if (this.visibility != visible) {
				this.visibility = visible;
				this.setEnabled(visible);
				Container parent = this.getParent();

> 				if (parent != null) {

Reviewer 2: Why do we repaint the parent (I understand it but maybe not the user)?

					parent.requestRender();
				}
			}
		}

		@Override
		public void render(GraphicsContext g) {
			if (this.visibility) {
				super.render(g);
			}
		}
	}

> Adding this to a button in the DemoWidget sample application yields these results:

Reviewer 2: How is created this container? What are its children? How do you change the visible state dynamically?
